docs/0003-hashed-n-tuple-trees.md

```diff
@@ -0,0 +1,260 @@
+  * **Default:** "toLower"
+* **Name**: `tupleSize`
+  * **Description:** Indicates the size of the segments (in characters) that the digest is split into
+  * **Type:** integer
+  * **Range:** 0,32
+  * **Default:** 3
+* **Name:** `numberOfTuples`
+  * **Description:** Indicates how many segments are used for path generation
+  * **Type:** integer
+  * **Range:** 0,32
+  * **Default:** 3
+* **Name:** `shortObjectRoot`
+  * **Description:** When true, indicates that the OCFL object root directory name should contain the remainder of the digest not used in the n-tuples segments
+  * **Type:** boolean
+  * **Default:** false
+
+### Details
+
+#### digestAlgorithm
+
+`digestAlgorithm` is defaulted to `sha256`,and it MUST either contain a digest algorithm that's [officially supported by the OCFL spec](https://ocfl.io/draft/spec/#digest-algorithms) or defined in a community extension. The specified algorithm is applied to OCFL object identifiers to produce hex encoded digest values that are then mapped to OCFL object root paths.
+
+#### caseMapping
+
+Hex numbers are case insensitive, however many filesystems are not. To avoid case related problems, all hex encoded digest values MUST be cased consistently. When `caseMapping` is set to `toLower`, the default, all letters in the digest MUST be converted to lowercase. When it's set to `toUpper`, all letters MUST be converted to uppercase.
+
+#### tupleSize
+
+`tupleSize` determines the number of digest characters to include in each tuple. The tuples are used as directory names. The default value is `3`, which means that each directory in the OCFL storage hierarchy could contain up to 4096 files. Increasing this value increases the maximum number of files per directory.
+
+If `tupleSize` is set to `0`, then no tuples are created and `numberOfTuples` MUST also equal `0`.
+
+The product of `tupleSize` and `numberOfTuples` MUST be less than or equal to the number of characters in the hex encoded digest.
+
+#### numberOfTuples
+
+`numberOfTuples` determines how many tuples to create from the digest. The tuples are used as directory names, and each successive directory is nested within the previous. The default value is `3`, which means that every OCFL object root will be 4 directories removed from the OCFL storage root, 3 tuple directories plus 1 encapsulation directory. Increasing this value increases the depth of the OCFL storage hierarchy.
+
+If `numberOfTuples` is set to `0`, then no tuples are created and `tupleSize` MUST also equal `0`.
+
+The product of `numberOfTuples` and `tupleSize` MUST be less than or equal to the number of characters in the hex encoded digest.
+
+#### shortObjectRoot
+
+The directory that immediately encapsulates an OCFL object MUST either be named using the entire digest or the remainder of the digest that was not used in a tuple. When `shortObjectRoot` is set to `false`, the default, the entire digest is used, and, when it's `true` only the previously unused remainder is used.
```

Comment:
I would like to have the object id (encoded as needed) as the encapsulating directory. A nice thing about that is that if I see a file path in a log (or anywhere), I can immediately know what the object id is, without having to look at the inventory.json. Is there another layout extension that would provide this? Is there a way to allow that in this extension?

Reply:
Using object ids as encapsulation directories introduces a slew of new problems and edge cases. At this point, I'm somewhat inclined to follow @neilsjefferies lead and try to keep the extensions focused on specific solutions, in which case, the layout your describing should have its own extension.

Out of curiosity, @bcail, what is the format of your ids? Are they fixed length? If so, perhaps @neilsjefferies's extension would work for you?

Reply:
Nope, our ids are not fixed length (they're simple Fedora-3-style namespace:int format - eg. bdr:1). It's possible in the future we'd start using some fixed-length ids, but we'd still have all the old ids (unless we map them to new ids).

I guess url-encoding the UTF8 bytes of the object id for the encapsulation directory name wouldn't work? I know that would make the length of the encapsulation directory name longer than the object id - are there other problems?

Reply:
It's possible to url encode object ids and use them as encapsulation directories. The primary issue is filename length. What is the behavior when the encoded directory name exceeds 255 characters?

If this behavior were added to this extension, how would it be defined? When I implemented something like what you're asking for in ocfl-java, I had single implementation and configuring all of its options was a bit of a mess. I'm not saying that it couldn't have been improved, but this is part of the reason that I've come around to favoring specialized extensions rather than exhaustive extensions -- they're a lot easier to write and configure.

For example, in this case, there could be a hypothetical extension `0005-hashed-n-tuple-url-encoded-layout` (or whatever) that behaves how you'd like it to and could simply be defined with the following configuration:

```json
{
    "digestAlgorithm": "sha256",
    "caseMapping": "toLower",
    "tupleSize": 3,
    "numberOfTuples": 3
}
```

I think this would work better than trying to capture all of these options in a single extension, at the cost of a little redundancy in the extension definitions.

Reply:
Or, this hypothetical extension could be something like `0005-hashed-n-tuple-id-layout`. It uses a digest of the id to construct the n-tuple hierarchy and the id itself as the encapsulation directory. The id can optionally be url encoded (no need to enable encoding if you know your ids will never contain problematic characters). Object ids that result in encapsulation directories that are longer than 255 characters would not be allowed. This is not an insignificant restriction if you don't have full control over your ids.

```json
{
    "digestAlgorithm": "sha256",
    "caseMapping": "toLower",
    "tupleSize": 3,
    "numberOfTuples": 3,
    "urlEncode": true
}
```

Reply:
OK, I can submit a PR for a new extension. Would you be open to supporting an extension like that in ocfl-java?

Reply:
For sure. I am intending to add support for future extensions, particularly layout extensions.